File: typing/src/main/java/org/navgurukul/typingguru/utils/Utility.java
package org.navgurukul.typingguru.utils;

import android.content.Context;
import android.hardware.usb.UsbManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.TimeUnit;

public class Utility {
    private static final String TAG = "Utility";
    public static final String TYPE_PRACTICE_TYPING = "practicetyping";
    private static Map<String, Integer> keyMap;
    static {
        keyMap = new HashMap<>();
        keyMap.put("a", 29);
        keyMap.put("s", 47);
        keyMap.put("d", 32);
        keyMap.put("f", 34);
        keyMap.put("g", 35);
        keyMap.put("h", 36);
        keyMap.put("j", 38);
        keyMap.put("k", 39);
        keyMap.put("l", 40);
        keyMap.put(";", 74);
        keyMap.put(" ", 62);
        keyMap.put("q", 45);
        keyMap.put("w", 51);
        keyMap.put("e", 33);
        keyMap.put("r", 46);
        keyMap.put("t", 48);
        keyMap.put("y", 53);
        keyMap.put("u", 49);
        keyMap.put("i", 37);
        keyMap.put("o", 43);
        keyMap.put("p", 44);
        keyMap.put("z", 54);
        keyMap.put("x", 52);
        keyMap.put("c", 31);
        keyMap.put("v", 50);
        keyMap.put("b", 30);
        keyMap.put("n", 42);
        keyMap.put("m", 41);
        keyMap.put(",", 55);
        keyMap.put(".", 56);
        keyMap.put("/", 76);
        keyMap.put("tab", 61);
    }

    public static Integer getKeyCodeByText(String text) {
        return keyMap.get(text);
    }

    public static int calculateWPM(long elapsedTime, List<String> list) {
        long second = elapsedTime/1000;
        int size = list.size();
        //wpm formula
        // (x characters/5)/1m = y wpm
        int wpm = (int) ((((double)size/5)/second)*60);
        return wpm;
    }

    public static int calculateWPM(long elapsedTime, ArrayList<java.util.ArrayList<String>> list) {
        long second = elapsedTime/1000;
        int size = 0;
        for (ArrayList<String> l : list) {
            size = size + l.size();
        }
        //wpm formula
        // (x characters/5)/1m = y wpm
        int wpm = (int) ((((double)size/5)/second)*60);
        return wpm;
    }

    public static ArrayList<String> generateRandomWordList(ArrayList<String> list) {
        Random r = new Random();
        ArrayList<String> result = new ArrayList<>();
        for (int i = 0; i < 8; i++) {
            String s = list.get(r.nextInt(list.size()));
            result.add(s);
        }
        return result;
    }

    public static String convertMinutesToMMSS(int second) {
        long millis = 1000*second;
        String ms = String.format("%02d:%02d",
                TimeUnit.MILLISECONDS.toMinutes(millis) - TimeUnit.HOURS.toMinutes(TimeUnit.MILLISECONDS.toHours(millis)),
                TimeUnit.MILLISECONDS.toSeconds(millis) - TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(millis)));
        return ms;
    }

    public static boolean isOtgSupported(Context context) {
        try {
            UsbManager usbManager = (UsbManager)context.getSystemService(Context.USB_SERVICE);
            Logger.d(TAG, "USB Manager : "+usbManager);
            if (usbManager == null) {
                return false;
            }
            boolean isOtgSupported = context.getPackageManager().hasSystemFeature("android.hardware.usb.host");
            Logger.d(TAG, "is Otg Supported :"+isOtgSupported);
            return isOtgSupported;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    public static ArrayList<String> getAlphabetList() {
        char c;
        ArrayList<String> result = new ArrayList<>();
        for(c = 'a'; c <= 'z'; ++c)
            result.add(String.valueOf(c));
        return result;
    }
}

File: typing/src/main/java/org/navgurukul/typingguru/utils/Utility.kt
package org.navgurukul.typingguru.utils

import android.content.Context
import android.hardware.usb.UsbManager
import timber.log.Timber
import java.util.Random
import java.util.concurrent.TimeUnit

class Utility {
    private var keyMap: HashMap<String, Int> = HashMap()
    val TYPE_PRACTICE_TYPING = "practicetyping"

    private val alphabetList: ArrayList<String> by lazy {
        ArrayList()
    }

    fun getAlphabets() : ArrayList<String> {
        var c = 'a'
        while (c <= 'z') {
            alphabetList.add(c.toString())
            ++c
        }
        return alphabetList;
    }

    fun getKeyCodeByText(text: String?): Int? {
        return keyMap!![text]
    }

    fun calculateWPM(elapsedTime: Long, list: List<String?>): Int {
        val second = elapsedTime / 1000
        val size = list.size
        //wpm formula
        // (x characters/5)/1m = y wpm
        return (size.toDouble() / 5 / second * 60).toInt()
    }

    fun calculateWPM(elapsedTime: Long, list: ArrayList<java.util.ArrayList<String>>): Int {
        val second = elapsedTime / 1000
        var size = 0
        for (l in list) {
            size = size + l.size
        }
        //wpm formula
        // (x characters/5)/1m = y wpm
        return (size.toDouble() / 5 / second * 60).toInt()
    }

    fun generateRandomWordList(list: ArrayList<String>): ArrayList<String> {
        val r = Random()
        val result = ArrayList<String>()
        for (i in 0..7) {
            val s = list[r.nextInt(list.size)]
            result.add(s)
        }
        return result
    }

    fun convertMinutesToMMSS(second: Int): String {
        val millis = 1000 * second.toLong()
        return String.format(
            "%02d:%02d",
            TimeUnit.MILLISECONDS.toMinutes(millis) - TimeUnit.HOURS.toMinutes(
                TimeUnit.MILLISECONDS.toHours(millis)
            ),
            TimeUnit.MILLISECONDS.toSeconds(millis) - TimeUnit.MINUTES.toSeconds(
                TimeUnit.MILLISECONDS.toMinutes(millis)
            )
        )
    }

    fun isOtgSupported(context: Context): Boolean {
        return try {
            val usbManager =
                context.getSystemService(Context.USB_SERVICE) as UsbManager
            Timber.d(
                "USB Manager : $usbManager"
            )
            if (usbManager == null) {
                return false
            }
            val isOtgSupported =
                context.packageManager.hasSystemFeature("android.hardware.usb.host")
            Timber.d(
                "is Otg Supported :$isOtgSupported"
            )
            isOtgSupported
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }



    init {
        keyMap["a"] = 29
        keyMap["s"] = 47
        keyMap["d"] = 32
        keyMap["f"] = 34
        keyMap["g"] = 35
        keyMap["h"] = 36
        keyMap["j"] = 38
        keyMap["k"] = 39
        keyMap["l"] = 40
        keyMap[";"] = 74
        keyMap[" "] = 62
        keyMap["q"] = 45
        keyMap["w"] = 51
        keyMap["e"] = 33
        keyMap["r"] = 46
        keyMap["t"] = 48
        keyMap["y"] = 53
        keyMap["u"] = 49
        keyMap["i"] = 37
        keyMap["o"] = 43
        keyMap["p"] = 44
        keyMap["z"] = 54
        keyMap["x"] = 52
        keyMap["c"] = 31
        keyMap["v"] = 50
        keyMap["b"] = 30
        keyMap["n"] = 42
        keyMap["m"] = 41
        keyMap[","] = 55
        keyMap["."] = 56
        keyMap["/"] = 76
        keyMap["tab"] = 61
        keyMap["1"] = 8
        keyMap["2"] = 9
        keyMap["3"] = 10
        keyMap["4"] = 11
        keyMap["5"] = 12
        keyMap["6"] = 13
        keyMap["7"] = 14
        keyMap["8"] = 15
        keyMap["9"] = 16
        keyMap["0"] = 7
    }
}